tools/reactive-controllers/src/PendingState.ts

```diff
@@ -0,0 +1,103 @@
+/*
+Copyright 2024 Adobe. All rights reserved.
+This file is licensed to you under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+
+import { html, ReactiveController, ReactiveElement, TemplateResult } from 'lit';
+import('@spectrum-web-components/progress-circle/sp-progress-circle.js');
+
+type PendingStateConfig = {
+    pending: () => boolean;
+};
+
+export interface HostWithPendingState extends ReactiveElement {
+    pendingLabel?: string;
+    pending?: boolean;
+}
+
+/**
+ * Represents a controller for managing the pending state of a reactive element.
+ *
+ * @template T - The type of the reactive element.
+ */
+export class PendingStateController<T extends HostWithPendingState>
+    implements ReactiveController
+{
+    /**
+     * The host element that this controller is attached to.
+     */
+    public host: T;
+
+    /**
+     * A function that returns a boolean indicating whether the host is in a pending state.
+     */
+    private pending: () => boolean;
+
+    /**
+     * A callback function that is called when the pending state changes.
+     *
+     * @param pending - A boolean indicating the new pending state.
+     */
+
+    /**
+     * Creates an instance of PendingStateController.
+     *
+     * @param host - The host element that this controller is attached to.
+     * @param config - The configuration object containing the pending function and onPendingChange callback.
+     */
+    constructor(host: T, { pending }: PendingStateConfig) {
+        this.host = host;
+        this.pending = pending;
+
+        this.host.addController(this);
+    }
+
+    /**
+     * Checks if the host is in a pending state.
+     *
+     * @returns A boolean indicating whether the host is in a pending state.
+     */
+    public isPending(): boolean {
+        return this.pending();
+    }
+
```

Comment:
I just realised since `pending` is a public property of our components, instead of line 68 `return this.pending()` you can do `return this.host.pending` and everything should still work the same. This means you can remove all the `this.pending()` instances from the controller and you don't need to set it up in the constructor either.

Reply:
Thank you for the suggestion. While it's true that accessing the pending property directly from the host (this.host.pending) would work, there are several reasons why retaining the current implementation with the this.pending() function is preferable:

1. Encapsulation and Flexibility:

Using the this.pending() function allows for better encapsulation. It provides a layer of abstraction between the controller and the host, making it easier to manage and modify the logic for determining the pending state without changing the controller's implementation. This encapsulation ensures that changes to the pending state logic do not directly impact the controller.

2. Custom Logic:

The pending function can encapsulate more complex logic for determining the pending state(if required in future). This flexibility allows for conditions or calculations that go beyond a simple property check. If the logic for determining the pending state evolves, it can be adjusted in the function without altering the controller.

3. Consistency and Predictability:

Using the this.pending() function ensures a consistent approach to checking the pending state across different components and contexts. It provides a single point of truth for the pending state logic, making the codebase easier to maintain and understand.

4. Avoiding Direct Property Access:

Directly accessing properties of the host from the controller can lead to tighter coupling between the controller and the host. This can make the code more fragile and harder to refactor. The function-based approach promotes a more modular and decoupled design.

Given these benefits, I believe it is better to retain the current implementation with the this.pending() function. It offers greater flexibility, maintainability, and encapsulation, which align with good software design principles.

Reply:
Yo... So I understand that you're trying to make everything flexible encapsulated etc. but accessing the pending property using `this.host.pending` isn't going to take away any of those things from us. It's just a different and faster way of doing the same thing if not the correct. In our other controllers we follow the same code structure where we use the host property getter to get the public properties instead of writing another layer of custom function over it as this unnecessarily creates extra code to maintain (might affect performance also) and doesn't provide clear benefit at all. 
TLDR; `this.host.pending` isn't taking away any of the flexibility, maintainability, encapsulation etc. Infact using `this.host` is the standard way of writing custom lit controllers for web components. 
Please go through the official docs [here](https://lit.dev/docs/composition/controllers/).

Reply:
Got it! Makes sense!
It just came to my mind that it will also reduce once function call as well.
Thanks for pointing this out!